swarm/bmt/bmt.go

```diff
@@ -382,83 +363,225 @@ func (h *Hasher) Write(b []byte) (int, error) {
 
 // Reset needs to be called before writing to the hasher
 func (h *Hasher) Reset() {
-	h.getTree()
+	h.releaseTree()
 }
 
-// Hasher implements the SwarmHash interface
+// methods needed to implement the SwarmHash interface
 
 // ResetWithLength needs to be called before writing to the hasher
 // the argument is supposed to be the byte slice binary representation of
 // the length of the data subsumed under the hash, i.e., span
 func (h *Hasher) ResetWithLength(span []byte) {
 	h.Reset()
-	h.bmt.span = span
+	h.getTree().span = span
 }
 
 // releaseTree gives back the Tree to the pool whereby it unlocks
 // it resets tree, segment and index
 func (h *Hasher) releaseTree() {
 	t := h.bmt
-	if t != nil {
-		t.cur = 0
+	if t == nil {
+		return
+	}
+	h.bmt = nil
+	go func() {
+		t.cursor = 0
 		t.offset = 0
 		t.span = nil
-		t.hash = nil
-		h.bmt = nil
 		t.section = make([]byte, h.pool.SegmentSize*2)
-		t.segment = make([]byte, h.pool.SegmentSize)
+		select {
+		case <-t.result:
+		default:
+		}
 		h.pool.release(t)
+	}()
+}
+
+// NewAsyncWriter extends Hasher with an interface for concurrent segment/section writes
+func (h *Hasher) NewAsyncWriter(double bool) *AsyncHasher {
+	secsize := h.pool.SegmentSize
+	if double {
+		secsize *= 2
+	}
+	write := func(i int, section []byte, final bool) {
+		h.writeSection(i, section, double, final)
 	}
+	return &AsyncHasher{
+		Hasher:  h,
+		double:  double,
+		secsize: secsize,
+		write:   write,
+	}
+}
+
+// SectionWriter is an asynchronous segment/section writer interface
```

Comment:
It appears that SectionWriter is defined but it is called AsyncWriter in comments. Do we need this interface, as it has only one implementation? If the intention is to have multiple implementations in the future, then it should be here (and consistently named), but if not, I do not see why it would exist.